VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol

```diff
@@ -0,0 +1,1098 @@
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function updateStakePerSlot(uint newVal) external payable onlyOwner override {
+        if (newVal == 0 || newVal == _slotInfo.STAKE_PER_SLOT) {
+            revert InvalidValue();
+        }
+
+        _slotInfo.STAKE_PER_SLOT = newVal;
+        emit UpdateStakePerSlot(newVal);
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function getSlotCountRange() external view returns(uint, uint) {
+        return (_slotInfo.MIN_SLOTS, _slotInfo.MAX_SLOTS);
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function updateMinSlotCount(uint minSlots) external payable onlyOwner override {
+        if (minSlots == 0 || minSlots == _slotInfo.MIN_SLOTS || minSlots > _slotInfo.MAX_SLOTS) {
+            revert InvalidValue();
+        }
+
+        _slotInfo.MIN_SLOTS = minSlots;
+        emit UpdateMinSlotCount(minSlots);
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function updateMaxSlotCount(uint maxSlots) external payable onlyOwner override {
+        if (maxSlots == 0 || maxSlots == _slotInfo.MAX_SLOTS || maxSlots < _slotInfo.MIN_SLOTS) {
+            revert InvalidValue();
+        }
+
+        _slotInfo.MAX_SLOTS = maxSlots;
+        emit UpdateMaxSlotCount(maxSlots);
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function updateTokenAddress(address newTokenAddress) external payable onlyOwner {
+        // Check the validity of the token address
+        {
+            if (vdaTokenAddress == newTokenAddress) {
+                revert InvalidTokenAddress();
+            }
+
+            assembly {
+                if iszero(newTokenAddress) {
+                    let ptr := mload(0x40)
+                    mstore(ptr, 0x1eb00b0600000000000000000000000000000000000000000000000000000000)
+                    revert(ptr, 0x4) //revert InvalidTokenAddress()
+                }
+            }
+        }
+
+        address oldAddress = vdaTokenAddress;
+        vdaTokenAddress = newTokenAddress;
+
+        emit UpdateTokenAddress(oldAddress, newTokenAddress);
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function getBalance(address didAddress) external view returns(uint) {
+        return _stakedTokenAmount[didAddress];
+    }
+
+    /**
+     * @notice Calculate the excess token amount for a DID address
+     * @dev Internal function used in `excessTokenAmount()` and `withdraw()` functions
+     * @param didAddress DID address
+     * @return uint Return 0 if staked amount is less than the required amount
+     */
+    function getExcessTokenAmount(address didAddress) internal view returns(uint) {
+        uint totalAmount;
+
+        uint nodeId = _didNodeId[didAddress];
+        if (nodeId != 0) {
+            totalAmount = requiredTokenAmount(_nodeMap[nodeId].slotCount);    
+        }
+
+        if (_stakedTokenAmount[didAddress] <= totalAmount) {
```

Comment:
I think we should return a negative value of there aren't enough tokens staked. That way node operators can know exactly the amount they need to stake to meet their requirements.

Reply:
That make sense.
Let me check on my side.

Reply:
Updated